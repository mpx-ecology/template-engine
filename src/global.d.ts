declare module '@mpxjs/webpack-plugin/lib/config'
declare module '@mpxjs/webpack-plugin/lib/utils/match-condition'
