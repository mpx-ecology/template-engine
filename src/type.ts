
export type ModeConfig = {
  typeExtMap: {
    json: string
    script: string
    template: string
    styles: string
  }
  tabBar: {
    customKey: string
    itemKey: string
    iconKey: string
    activeIconKey: string
  }
  event: {
    parseEvent(attr: any):
      | {
          prefix: string
          eventName: string
          modifier: string
          custom: boolean
        }
      | undefined
    getEvent(eventName: any, prefix?: string): string
    defaultModelProp: string
    defaultModelEvent: string
    defaultModelValuePath: string
    shallowStringify(obj: any): string
  }
  wxs: {
    tag: string
    module: string
    src: string
    ext: string
    templatePrefix: string
  }
  directive: {
    if: string
    elseif: string
    else: string
    model: string
    modelProp: string
    modelEvent: string
    modelValuePath: string
    modelFilter: string
    for: string
    forIndex: string
    forItem: string
    key: string
    dynamicClass: string
    dynamicStyle: string
    ref: string
    show: string
  }
}
