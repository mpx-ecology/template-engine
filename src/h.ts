import { NodeType } from './shortcuts'

export type VNode = {
  tag: string
  attrs?: Record<string, string>
  children?: (VNode | string)[]
}

export function h(
  tag: VNode['tag'],
  attrs: VNode['attrs'],
  children: VNode['children']
) {
  return {
    tag,
    attrs,
    children
  }
}

function buildAttrString(vnode: VNode) {
  const result: string[] = []
  Object.entries(vnode.attrs || {}).forEach(function ([key, value]) {
    if (key.startsWith('__')) return
    result.push(value ? key + '="' + value + '"' : key)
  })
  return result.join(' ')
}

export function toHtml(vnode?: VNode | VNode[] | string): string {
  if (Array.isArray(vnode)) return vnode.map((v) => toHtml(v)).join('')
  if (typeof vnode === 'undefined' || vnode === null) return ''
  if (typeof vnode === 'string') return vnode

  const tagName = vnode.tag
  const attributes = buildAttrString(vnode)

  const html: string[] = []

  // Open tag
  html.push('<' + tagName)
  if (attributes.length) html.push(' ' + attributes)
  html.push(vnode.children?.length ? '>' : '/>')
  // children
  vnode.children?.forEach(function (child) {
    html.push(toHtml(child))
  })
  // End Tag
  vnode.children?.length && html.push('</' + tagName + '>')
  return html.join('')
}

export function createNodeGenerator(type: string) {
  return function createNode(
    attrs: VNode['attrs'],
    children?: VNode['children']
  ) {
    return h(type, attrs, children)
  }
}

export const template = createNodeGenerator(NodeType.Template)
export const view = createNodeGenerator(NodeType.View)
export const block = createNodeGenerator(NodeType.Block)
export const element = createNodeGenerator(NodeType.Element)
export const wxs = createNodeGenerator(NodeType.Wxs)
export const component = (
  nodeType: string,
  attrs: VNode['attrs'],
  children?: VNode['children']
) => {
  switch (nodeType) {
    case NodeType.Slot:
    case NodeType.CatchView:
    case NodeType.StaticView:
    case NodeType.PureView:
      nodeType = NodeType.View
      break
    case NodeType.StaticText:
      nodeType = NodeType.Text
      break
    case NodeType.StaticImage:
      nodeType = NodeType.Image
      break
  }
  return createNodeGenerator(nodeType)(attrs, children)
}
