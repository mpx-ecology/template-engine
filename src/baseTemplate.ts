import {
  focusComponents,
  baseComponents,
  nestElements,
  singleQuote,
  voidElements
} from './components'
import { component, template, view, element, block, VNode } from './h'
import { isBooleanStringLiteral, isNumber } from './is'
import { getOptimizedComponentInfo } from './optimizer'
import {
  ContextData,
  Context,
  NodeType,
  Prop,
  ContextName,
  XsUtils,
  MpxGLobal
} from './shortcuts'
import { ModeConfig } from './type'
import {
  flatArrWithObjectToObject,
  hasOwn,
  isCustomEvent,
  processAddComponentRootView,
  toCamelCase,
  toDashed
} from './utils'
import config from '@mpxjs/webpack-plugin/lib/config'

export interface Component {
  nodeType: string
  attrs: Attributes
}

export type InputComponentConfig = {
  nodeType: string
  attrs?: (string | Attributes)[] | Attributes
}

export type InputComponentsOptions =
  | (InputComponentConfig | string)[]
  | Record<string, Record<string, string> | string[]>

export type InputOptions = {
  baseComponents?: InputComponentsOptions // view
  normalComponents?: InputComponentsOptions // .mpx
  runtimeComponents?: InputComponentsOptions // .runtime.mpx
  inlineSlot?: boolean
  templateCompilerConfig?: any
  proxyNormalComponentAttrs?: boolean
}

export type BuildComponentOptions = Record<string, Record<string, string>>

export interface BuildOptions {
  baseComponents: Component[] // view
  normalComponents: Component[] // .mpx
  runtimeComponents: Component[] // .runtime.mpx
  inlineSlot?: boolean
  templateCompilerConfig?: any
  proxyNormalComponentAttrs?: boolean
}

export type Attributes = Record<string, string>

export type SUPPORT_MODE =
  | 'wx'
  | 'ali'
  | 'swan'
  | 'qq'
  | 'tt'
  | 'dd'
  | 'web'
  | 'tenon'

export class BaseTemplate {
  protected exportExpr = 'module.exports ='
  protected isSupportRecursive = false
  protected buildOptions: BuildOptions = {
    baseComponents: [],
    normalComponents: [],
    runtimeComponents: []
  }
  /** 组件列表 */
  public baseComponents = baseComponents
  /** 可以 focus 聚焦的组件 */
  public focusComponents: Set<string> = focusComponents
  /** 不需要渲染子节点的元素 */
  public voidElements: Set<string> = voidElements
  /** 可以递归调用自身的组件 */
  public nestElements: Map<string, number> = nestElements

  modeConfig: ModeConfig

  mode = 'wx'

  private _baseLevel = 8

  set baseLevel(lv) {
    this._baseLevel = lv
  }

  get baseLevel() {
    return this._baseLevel
  }

  get Directive() {
    return this.modeConfig.directive
  }

  constructor(mode?: SUPPORT_MODE) {
    this.mode = mode || 'wx'
    this.modeConfig = config[this.mode]
  }

  protected isEvent(event: string) {
    const res = this.modeConfig.event.parseEvent(event)
    if (res) res.custom = isCustomEvent(res.eventName)
    return res
  }

  /**
   * 获取vnode节点属性
   * @param value -
   * @returns
   */
  protected getContext(value?: string, dynamic = true) {
    const v = value ? ContextName + '.' + value : ContextName
    return dynamic ? this.dynamicValue(v) : v
  }

  /**
   * 获取vnode节点data数据
   * @param value -
   * @returns
   */
  protected getContextData(value?: string, nodeType?: string): string {
    const baseComponent = nodeType && this.baseComponents[nodeType]
    if (baseComponent && value && baseComponent[value]) {
      return `${this.getXsUtils(XsUtils.B)}(${this.getContextData(value)}, ${baseComponent[value]})`
    }
    return ContextName + `.${Context.Data}` + (value ? `.${value}` : '')
  }

  protected getXsUtils(name: string) {
    return `${XsUtils.Module}.${name}`
  }

  protected transContext(keymap: string = ContextName) {
    return `${ContextName}:${keymap}`
  }

  // 模板动态值: {{ value }}
  protected dynamicValue(s: string) {
    return `{{${s}}}`
  }

  protected getTplName(name: string, level?: string | number) {
    return `t_${level === undefined ? `\${level}` : level}_` + name
  }

  protected getNextTplName(name: string, level?: string | number) {
    return this.getTplName(name, level === undefined ? `\${nextLevel}` : level)
  }

  private getChildren(comp: Component) {
    if (this.voidElements.has(comp.nodeType)) return
    return [
      block(
        {
          [this.Directive.for]: this.getContext(Context.Children),
          [this.Directive.key]: 'index'
        },
        [
          template({
            [Prop.Is]: this.getNextTplName(NodeType.Container),
            [Prop.Data]: this.dynamicValue(this.transContext('item'))
          })
        ]
      )
    ]
  }

  private buildAttribute(attrs: Attributes) {
    return Object.keys(attrs)
      .filter((v) => v !== Prop.RawTag)
      .reduce(
        (res, k) => {
          res[k] = this.isEvent(k) ? attrs[k] : this.dynamicValue(attrs[k])
          return res
        },
        {} as Record<string, string>
      )
  }

  // 递归生成基础模板，被注入到 base.wxml 当中的文本内容
  protected buildBaseTemplate() {
    return template({ [Prop.Name]: 'mpx_tmpl' }, [
      element({
        [Prop.R]: this.dynamicValue('r'),
        [this.Directive.if]: this.dynamicValue('r')
      })
    ])
  }

  // 构建组件的模板
  protected buildComponentTemplate(comp: Component) {
    return this.focusComponents.has(comp.nodeType)
      ? this.buildFocusComponentTemplate(comp)
      : this.buildStandardComponentTemplate(comp)
  }

  protected buildComponentsTemplate(buildOptions: BuildOptions) {
    const res: VNode[] = []
    buildOptions.baseComponents.forEach((component) => {
      res.push(...this.buildComponentTemplate(component))
    })
    return res
  }

  protected buildFocusComponentTemplate(comp: Component) {
    const child = this.getChildren(comp)

    const fAttrs = comp.attrs
    const bAttrs = { ...comp.attrs }
    delete bAttrs.focus

    const fName = this.getTplName(comp.nodeType) + '_f'
    const bName = this.getTplName(comp.nodeType) + '_b'

    return [
      template({ name: this.getTplName(comp.nodeType) }, [
        template({
          [Prop.Is]: this.dynamicValue(
            `${this.getContext('focus', false)} ? ${fName} : ${bName}`
          ),
          [Prop.Data]: this.dynamicValue(
            `${this.transContext()}${child ? ',cid:cid' : ''}`
          )
        })
      ]),
      template({ [Prop.Name]: fName }, [
        component(comp.nodeType, this.buildAttribute(fAttrs), child)
      ]),
      template({ [Prop.Name]: bName }, [
        component(comp.nodeType, this.buildAttribute(bAttrs), child)
      ])
    ]
  }

  protected buildStandardComponentTemplate(comp: Component) {
    const res: VNode[] = []
    const nodeType = comp.nodeType
    const children = this.getChildren(comp)
    const normalTemplate = template(
      { [Prop.Name]: this.getTplName(nodeType) },
      [
        component(
          comp.attrs.rawTag || nodeType,
          this.buildAttribute(comp.attrs),
          children
        )
      ]
    )
    res.push(normalTemplate)
    return res
  }

  protected buildPlainTextTemplate() {
    return template(
      {
        [Prop.Name]: this.getTplName(NodeType.Text),
        [Prop.Data]: this.dynamicValue(this.transContext())
      },
      [block({}, [this.getContext(Context.Text)])]
    )
  }

  protected buildRunTimeComponentsTemplate(buildOptions: BuildOptions) {
    const nodes: VNode[] = []
    buildOptions.runtimeComponents.forEach((v) => {
      nodes.push(
        template({ [Prop.Name]: this.getTplName(v.nodeType) }, [
          component(v.nodeType, this.buildAttribute(v.attrs))
        ])
      )
    })
    return nodes
  }

  // 包括 custom-wrapper、原生的小程序组件、第三方(例如 vant)的组件
  protected buildThirdPartyTemplate(buildOptions: BuildOptions) {
    if (buildOptions.inlineSlot)
      return this.buildThirdPartyTemplateInlineSlot(buildOptions)
    const nodes: VNode[] = []
    if (Object.keys(buildOptions.normalComponents).length) {
      nodes.push(
        this.buildSlotTemplate(
          template({
            [Prop.Is]: this.getNextTplName(NodeType.Container),
            [Prop.Data]: this.dynamicValue(this.transContext())
          })
        )
      )
      buildOptions.normalComponents.forEach((v) => {
        let componentNode = component(
          v.nodeType,
          this.buildAttribute(v.attrs),
          [
            block(
              {
                [this.Directive.for]: this.getContext(Context.Children),
                [this.Directive.forItem]: ContextName,
                [this.Directive.key]: 'index'
              },
              [
                template({
                  [Prop.Is]: this.getTplName(NodeType.Slot),
                  [Prop.Data]: this.dynamicValue(this.transContext())
                })
              ]
            )
          ]
        )
        if (this.buildOptions.proxyNormalComponentAttrs) {
          componentNode = processAddComponentRootView(componentNode)
        }
        nodes.push(
          template({ name: this.getTplName(v.nodeType) }, [componentNode])
        )
      })
    }

    nodes.push(...this.buildRunTimeComponentsTemplate(buildOptions))

    return nodes
  }

  protected buildThirdPartyTemplateInlineSlot(buildOptions: BuildOptions) {
    const nodes: VNode[] = []

    buildOptions.normalComponents.forEach((v) => {
      const child = [
        template({
          [Prop.Is]: this.getNextTplName(NodeType.Container),
          [Prop.Data]: this.getContext()
        })
      ]

      let componentNode = component(v.nodeType, this.buildAttribute(v.attrs), [
        block(
          {
            [this.Directive.for]: this.getContext(Context.Children),
            [this.Directive.key]: 'index',
            [this.Directive.forItem]: ContextName
          },
          [
            block(
              {
                [this.Directive.if]: this.dynamicValue(
                  this.getContextData(ContextData.Slot)
                )
              },
              [
                view(
                  {
                    slot: this.dynamicValue(
                      this.getContextData(ContextData.Slot)
                    )
                  },
                  child
                )
              ]
            ),
            block({ [this.Directive.else]: '' }, child)
          ]
        )
      ])

      if (this.buildOptions.proxyNormalComponentAttrs) {
        componentNode = processAddComponentRootView(componentNode)
      }

      nodes.push(
        template({ name: this.getTplName(v.nodeType) }, [componentNode])
      )
    })

    nodes.push(...this.buildRunTimeComponentsTemplate(buildOptions))

    return nodes
  }

  protected buildContainerTemplate(restart = false) {
    const tmpl: VNode[] = []
    if (restart) {
      tmpl.push(
        block(
          {
            [this.Directive.if]: this.dynamicValue(
              `${this.getContext(Context.NodeType, false)} === '${NodeType.Text}'`
            )
          },
          [
            template({
              [Prop.Is]: this.getTplName(NodeType.Text, '0'),
              [Prop.Data]: this.dynamicValue(this.transContext())
            })
          ]
        ),
        block({ [this.Directive.else]: '' }, [
          element({ r: this.getContext() })
        ])
      )
    } else {
      tmpl.push(
        template({
          [Prop.Is]: this.getTplName(this.getContext(Context.NodeType)),
          [Prop.Data]: this.dynamicValue(this.transContext())
        })
      )
    }

    return template({ [Prop.Name]: this.getTplName(NodeType.Container) }, tmpl)
  }

  protected buildSlotTemplate(child: VNode) {
    return template({ [Prop.Name]: this.getTplName(NodeType.Slot) }, [
      block(
        {
          [this.Directive.if]: this.dynamicValue(
            this.getContextData(ContextData.Slot)
          )
        },
        [
          view(
            { slot: this.dynamicValue(this.getContextData(ContextData.Slot)) },
            [child]
          )
        ]
      ),
      block({ [this.Directive.else]: '' }, [child])
    ])
  }

  protected genBaseComponentsAttrs(attrs: Attributes, nodeType: NodeType) {
    const newAttrs: Attributes = Object.create(null)
    for (const prop in attrs) {
      if (hasOwn(attrs, prop)) {
        let propValue = attrs[prop]
        // 事件绑定
        const eventParsed = this.isEvent(prop)
        if (eventParsed) {
          propValue = this.dynamicValue(this.getXsUtils(XsUtils.Invoke))
          newAttrs[Prop.DataEventConfig] = this.getContextData(
            toCamelCase(Prop.DataEventConfig)
          )
          newAttrs[Prop.DataMpxUid] = this.getContextData(ContextData.Uid)
        } else if (prop === Prop.RawTag) {
          // do nothing
        } else if (propValue === '') {
          // <button primary></button> 单属性值的处理
          propValue = this.getContextData(toCamelCase(prop), nodeType)
        } else if (isBooleanStringLiteral(propValue) || isNumber(+propValue)) {
          const dValue = this.getContextData(toCamelCase(prop), nodeType)
          propValue = `${dValue}===undefined?${propValue}:${dValue}`
        } else {
          const dValue = this.getContextData(toCamelCase(prop), nodeType)
          propValue = `${dValue}||${propValue || singleQuote('')}`
        }
        newAttrs[prop] = propValue
      }
    }
    // 添加默认的属性，events
    if (nodeType !== NodeType.Block) {
      if (newAttrs[Prop.Style] || newAttrs[Prop.Class]) {
        newAttrs[Prop.Style] = this.getContextData(ContextData.Style)
      }
      if (attrs[Prop.Class]) {
        newAttrs[Prop.Class] = this.getContextData(ContextData.Class)
      }
    }
    // swiper-item 去除 style 配置
    if (nodeType === NodeType.SwiperItem) {
      delete newAttrs[Prop.Style]
    }
    return newAttrs
  }

  // 标准化内置component
  protected createBaseComponents(components: Component[]): Component[] {
    const result: Component[] = []
    components.forEach((v) => {
      const nodeType = toDashed(v.nodeType) as NodeType
      if (nodeType === NodeType.Slot) {
        return result.push({
          nodeType,
          attrs: {
            [Prop.Slot]: this.getContextData(ContextData.Name)
          }
        })
      }
      // 获取优化后节点及属性
      const optimizedNodeInfo = getOptimizedComponentInfo(
        {
          nodeType,
          attrs: v.attrs
        },
        this.mode
      )
      const attrs: Attributes = this.genBaseComponentsAttrs(
        optimizedNodeInfo?.attrs || v.attrs,
        nodeType
      )
      if (optimizedNodeInfo?.optimizeTag) {
        const { nodeType: optimizedNodeType } = optimizedNodeInfo
        // view
        if (optimizedNodeType === nodeType) {
          result.push({
            nodeType,
            attrs
          })
        }
        // static-view
        if (
          optimizedNodeType === nodeType ||
          optimizedNodeType.startsWith('static')
        ) {
          result.push({
            nodeType: `static-${nodeType}`,
            attrs: this.genBaseComponentsAttrs(
              {
                ...optimizedNodeInfo.staticAttrs,
                rawTag: nodeType
              },
              nodeType
            )
          })
        }
        // pure
        result.push({
          nodeType: `pure-${nodeType}`,
          attrs: this.genBaseComponentsAttrs(
            {
              ...optimizedNodeInfo.pureAttrs,
              rawTag: nodeType
            },
            nodeType
          )
        })
      } else {
        result.push({
          nodeType,
          attrs
        })
      }
    })
    return result
  }

  // 标准化自定义component
  protected createNormalComponents(components: Component[]): Component[] {
    const result: Component[] = []
    components.forEach((v) => {
      const optimizedNodeInfo = getOptimizedComponentInfo(
        {
          nodeType: v.nodeType,
          attrs: v.attrs
        },
        this.mode
      )
      const attrs = { ...(optimizedNodeInfo?.attrs || v.attrs) }
      for (const prop in attrs) {
        const eventParsed = this.isEvent(prop)
        if (eventParsed) {
          attrs[prop] = eventParsed.custom
            ? this.dynamicValue(this.getXsUtils(XsUtils.Invoke))
            : MpxGLobal.Invoke
          attrs[Prop.DataEventConfig] = this.getContextData(
            toCamelCase(Prop.DataEventConfig)
          )
          attrs[Prop.DataMpxUid] = this.getContextData(ContextData.Uid)
        } else if (prop === Prop.MpxAttrs) {
          attrs[prop] = this.getContextData()
        } else if (prop === Prop.MpxShow) {
          const dValue = this.getContextData(ContextData.MpxShow)
          attrs[prop] = `${dValue} === undefined ? true : ${dValue}`
        } else if (prop === Prop.Class && !attrs[Prop.Style]) {
          attrs[Prop.Style] = this.getContextData(ContextData.Style)
          attrs[Prop.Class] = this.getContextData(ContextData.Class)
        } else {
          attrs[prop] = this.getContextData(toCamelCase(prop))
        }
      }
      result.push({
        nodeType: v.nodeType,
        attrs: attrs
      })
    })

    return result
  }

  protected normalizeInputComponentOptions(
    componentOptions?: InputComponentsOptions
  ): Component[] {
    if (!componentOptions) return []
    const res: Component[] = []
    if (Array.isArray(componentOptions)) {
      componentOptions.forEach((componentConfig) => {
        if (typeof componentConfig === 'string') {
          res.push({ nodeType: componentConfig, attrs: {} })
        } else {
          res.push({
            nodeType: componentConfig.nodeType,
            attrs: flatArrWithObjectToObject(componentConfig.attrs || [])
          })
        }
      })
    } else {
      // { view: {} }
      Object.keys(componentOptions).forEach((key: string) => {
        res.push({
          nodeType: key,
          attrs: flatArrWithObjectToObject(componentOptions[key])
        })
      })
    }
    return res
  }

  protected normalizeInputOptions(inputOptions: InputOptions) {
    const baseComponents = this.normalizeInputComponentOptions(
      inputOptions.baseComponents
    )
    const runtimeComponents = this.normalizeInputComponentOptions(
      inputOptions.runtimeComponents
    )
    const normalComponents = this.normalizeInputComponentOptions(
      inputOptions.normalComponents
    )
    this.buildOptions = {
      baseComponents: this.createBaseComponents(baseComponents),
      runtimeComponents: this.createNormalComponents(runtimeComponents),
      normalComponents: this.createNormalComponents(normalComponents),
      inlineSlot: inputOptions.inlineSlot || this.mode === 'ali' ? true : false,
      templateCompilerConfig: inputOptions.templateCompilerConfig,
      proxyNormalComponentAttrs: inputOptions.proxyNormalComponentAttrs
    }
  }
}
