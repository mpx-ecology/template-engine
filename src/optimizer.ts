import { InputComponentConfig } from './baseTemplate'
import { NodeType, Prop } from './shortcuts'
import { ModeConfig } from './type'
import { flatArrWithObjectToObject, isPureAttr } from './utils'

export const OPTIMIZE_TAG_NODES = [
  NodeType.View,
  NodeType.Text,
  NodeType.Image
] as string[]

// eslint-disable-next-line @typescript-eslint/no-var-requires
const config = require('@mpxjs/webpack-plugin/lib/config')

export function getOptimizedComponentInfo(
  c: InputComponentConfig,
  mode = 'wx'
):
  | {
      nodeType: string
      attrs: Record<string, string>
      staticAttrs: Record<string, string>
      pureAttrs: Record<string, string>,
      optimizeTag: boolean
    }
  | undefined {
  let nodeType = c.nodeType
  if (!c.attrs) return
  const modeConfig: ModeConfig = config[mode]
  const originAttrs = flatArrWithObjectToObject(c.attrs)
  const attrs: Record<string, string> = {}
  const staticAttrs: Record<string, string> = {}
  const pureAttrs: Record<string, string> = {}
  let hasEvent = false // 有事件
  let hasStaticAttr = false // 有非pure属性
  Object.keys(originAttrs).forEach((name) => {
    if (name === Prop.DataEventConfig || name === Prop.DataMpxUid) return
    const value = originAttrs[name]
    const eventInfo = modeConfig.event.parseEvent(name)
    if (eventInfo) {
      hasEvent = true
      const event = modeConfig.event.getEvent(
        eventInfo.eventName,
        eventInfo.prefix
      )
      if (eventInfo.prefix === 'catch') {
        delete attrs[name]
        const bindEvent = modeConfig.event.getEvent(
          eventInfo.eventName,
          mode === 'ali' ? 'on' : 'bind'
        )
        // 替换catch=>bind
        attrs[bindEvent] = ''
      } else {
        // 去重
        attrs[event] = value
      }
    } else if (isPureAttr(name)) {
      attrs[name] = value
      staticAttrs[name] = value
      pureAttrs[name] = value
    } else {
      attrs[name] = value
      staticAttrs[name] = value
      hasStaticAttr = true
    }
  })

  if(OPTIMIZE_TAG_NODES.includes(c.nodeType)) {
    if (!hasStaticAttr && !hasEvent) {
      nodeType = `pure-${c.nodeType}`
    } else if (!hasEvent) {
      nodeType = `static-${c.nodeType}`
    }
  }

  return {
    nodeType,
    optimizeTag: OPTIMIZE_TAG_NODES.includes(c.nodeType),
    attrs,
    staticAttrs,
    pureAttrs
  }
}
