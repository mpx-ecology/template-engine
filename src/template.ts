/**
 * Fork from \@tarojs/shared for generating base template
 *
 * 这里我们需要关心的小程序种类有两类：
 * 1. 模板递归：
 *  - 支持：tmpl0 套 tmpl0
 *  - 不支持：这就使得我们必须生成多级的模板，tmpl0 套 tmpl1，tmpl1 套 tmpl2……
 *           直到超过阈值 N (N = config.miniapp.baseLevel) tmplN 会套组件 comp，组件 comp 重新再套 tmpl0。
 * 2. 小程序脚本语言（wxs, sjs, etc...）：
 *  - 支持：可以在模板使用函数缩减模板大小或提高性能（存疑），例如判断一个值是不是假值（falsy value）。
 *         将来或许会把数据序列化^1 的操作也放到小程序脚本语言里。
 *  - 不支持：使用纯 *xml 语法
 *
 */

import { InputOptions, BaseTemplate, SUPPORT_MODE } from './baseTemplate'
import { toHtml } from './h'

export { InputOptions }

export class RecursiveTemplate extends BaseTemplate {
  isSupportRecursive = true

  public buildTemplate = (inputOptions: InputOptions) => {
    this.normalizeInputOptions(inputOptions)
    const code = `
      const nextLevel = 0;
      return \`${toHtml([
        this.buildBaseTemplate(),
        ...this.buildComponentsTemplate(this.buildOptions),
        this.buildPlainTextTemplate(),
        ...this.buildThirdPartyTemplate(this.buildOptions),
        this.buildContainerTemplate()
      ])}\`
    `
    const template = new Function('level', code)(0)
    return template
  }
}

export class UnRecursiveTemplate extends BaseTemplate {
  isSupportRecursive = false

  public buildTemplate = (inputOptions: InputOptions) => {
    this.normalizeInputOptions(inputOptions)
    const buildFloor = this.createFloorBuilder()
    const template = new Array(this.baseLevel)
      .fill('')
      .reduce(
        (a, _, i) => a + buildFloor(i),
        `${toHtml([this.buildBaseTemplate()])}`
      )
    return template
  }

  protected createFloorBuilder() {
    const code = `
    const nextLevel = level + 1;
    const restart = ${this.baseLevel} === nextLevel;
    if (restart) return \`${toHtml(this.buildContainerTemplate(true))}\`;
    return \`${toHtml([
      ...this.buildComponentsTemplate(this.buildOptions),
      this.buildPlainTextTemplate(),
      ...this.buildThirdPartyTemplate(this.buildOptions),
      this.buildContainerTemplate()
    ])}\`
    `
    return new Function('level', code)
  }
}

export function createTemplateEngine(mode?: SUPPORT_MODE) {
  if (mode === 'ali') return new RecursiveTemplate(mode)
  return new UnRecursiveTemplate(mode)
}

export function createSetupTemplate() {
  return `<template is="t_0_container" data="{{ i: r }}" wx:if="{{r && r.nt}}"></template>`
}

export function createSetupWxsTemplate() {
  return `<wxs module="xs" src="~@mpxjs/template-engine/dist/utils.wxs"/>`
}

export function createCustomElementTemplate(prefix = '') {
  return `<template>${prefix ? '\n  ' + prefix : ''}
  ${createSetupWxsTemplate()}
</template>

<script>
// element 里面需要承载所有需要走运行时渲染的组件
import { createComponent } from '@mpxjs/core'

createComponent({
  options: {
    addGlobalClass: true,
    styleIsolation: 'shared',
    // todo: 超过基础模板的层数，virtualHost 置为 true，避免样式规则失效
    virtualHost: true
  },
  properties: {
    r: { // vdom 数据
      type: Object,
      value: {
        nt: 'block'
      }
    }
  },
  data: {
    // 运行时组件的标识
    mpxCustomElement: true
  },
  computed: {
    vnodeData () {
      const data = this.r.d || {}
      return data
    },
    // 当前组件的 uid
    uid () {
      return this.vnodeData.uid
    }
  }
})
</script>

<style>

</style>

<script type="application/json">
{
  "component": true,
  "usingComponents": {}
}
</script>`
}
