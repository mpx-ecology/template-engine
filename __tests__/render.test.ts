import fs from 'fs'
import simulate from 'miniprogram-simulate'
import { describe, expect, test } from 'vitest'
import { InputOptions, createTemplateEngine } from '../src/template'
import path from 'path'

function initTemplateComponent(componentConfig: InputOptions) {
  const templateEngine = createTemplateEngine('wx')
  fs.writeFileSync(
    path.resolve('./__tests__/components/dynamic/index.wxml'),
    [
      `<template is="t_0_container" data="{{ i: r }}" wx:if="{{ r && r.nt }}"></template>`,
      templateEngine.buildTemplate(componentConfig)
    ].join('\n'),
    {
      encoding: 'utf-8'
    }
  )
}

function renderComponent(vNode) {
  const id = simulate.load(path.resolve('./__tests__/components/dynamic/index'))
  const comp = simulate.render(id, {
    r: vNode
  })
  return comp
}

describe('render', () => {
  test('render', () => {
    initTemplateComponent({
      baseComponents: [
        {
          nodeType: 'view',
          attrs: {
            class: '',
            style: '',
            'bind:tap': '',
            'data-eventconfigs': ''
          }
        },
        { nodeType: 'block', attrs: { style: '' } },
        {
          nodeType: 'image',
          attrs: { src: '', openImageRetry: '', class: '', style: '', mode: '' }
        },
        {
          nodeType: 'swiper-item',
          attrs: { class: '', 'bind:tap': '', 'data-eventconfigs': '' }
        },
        {
          nodeType: 'swiper',
          attrs: {
            id: '',
            autoplay: '',
            interval: '',
            circular: '',
            class: '',
            'bind:change': '',
            'data-eventconfigs': ''
          }
        },
        { nodeType: 'text', attrs: { class: '' } },
        {
          nodeType: 'd783097f6',
          attrs: {
            class: '',
            'catch:tap': '',
            'data-eventconfigs': '',
            rawTag: 'view'
          }
        },
        { nodeType: 'button', attrs: {} }
      ],
      normalComponents: [{
        nodeType: 'm5f9d5dc4',
        attrs: {
          text: '',
          rules: '',
          class: ''
        }
      }]
    })

    const comp = renderComponent({
      nt: 'view',
      d: {
        uid: 2,
        style: 'color: red',
        a: '1'
      },
      c: [
        {
          nt: 'swiper',
          d: {
            uid: 3
          },
          c: []
        },
        {
          nt: 'button',
          d: {
            uid: 4,
            type: 'primary',
            style: 'color: red'
          }
        },
        {
          nt: 'm5f9d5dc4',
          d: {
            uid: 4,
            type: 'primary',
            style: 'color: red'
          },
          c: [
            {
              d: {
                slot: 'test',
                style: 'color: red',
                class: 'slot-view-class'
              },
              nt: 'view'
            }
          ]
        }
      ],
      scopeProcessed: true
    })

    expect(comp.toJSON()).toMatchSnapshot()
  })
})
