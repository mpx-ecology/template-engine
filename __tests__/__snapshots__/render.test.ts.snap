// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`render > render 1`] = `
{
  "attrs": [],
  "children": [
    {
      "attrs": [
        {
          "name": "class",
          "value": "",
        },
        {
          "name": "data-eventconfigs",
          "value": "",
        },
        {
          "name": "data-mpxuid",
          "value": 2,
        },
        {
          "name": "style",
          "value": "color: red",
        },
      ],
      "children": [
        {
          "attrs": [
            {
              "name": "autoplay",
              "value": "",
            },
            {
              "name": "circular",
              "value": "",
            },
            {
              "name": "class",
              "value": "",
            },
            {
              "name": "data-eventconfigs",
              "value": "",
            },
            {
              "name": "data-mpxuid",
              "value": 3,
            },
            {
              "name": "id",
              "value": "",
            },
            {
              "name": "interval",
              "value": "",
            },
            {
              "name": "style",
              "value": "",
            },
          ],
          "children": [],
          "event": {
            "change": {
              "handler": "",
              "isCapture": false,
              "isCatch": false,
              "isMutated": false,
              "name": "change",
            },
          },
          "tagName": "wx-swiper",
        },
        {
          "attrs": [],
          "children": [],
          "event": {},
          "tagName": "wx-button",
        },
        {
          "attrs": [
            {
              "name": "class",
              "value": "",
            },
            {
              "name": "rules",
              "value": "",
            },
            {
              "name": "style",
              "value": "color: red",
            },
            {
              "name": "text",
              "value": "",
            },
          ],
          "children": [
            {
              "attrs": [
                {
                  "name": "slot",
                  "value": "test",
                },
              ],
              "children": [
                {
                  "attrs": [
                    {
                      "name": "class",
                      "value": "slot-view-class",
                    },
                    {
                      "name": "data-eventconfigs",
                      "value": "",
                    },
                    {
                      "name": "data-mpxuid",
                      "value": "",
                    },
                    {
                      "name": "style",
                      "value": "color: red",
                    },
                  ],
                  "children": [],
                  "event": {
                    "tap": {
                      "handler": "",
                      "isCapture": false,
                      "isCatch": false,
                      "isMutated": false,
                      "name": "tap",
                    },
                  },
                  "tagName": "wx-view",
                },
              ],
              "event": {},
              "tagName": "wx-view",
            },
          ],
          "event": {},
          "tagName": "m5f9d5dc4",
        },
      ],
      "event": {
        "tap": {
          "handler": "",
          "isCapture": false,
          "isCatch": false,
          "isMutated": false,
          "name": "tap",
        },
      },
      "tagName": "wx-view",
    },
  ],
  "event": {},
  "tagName": "main",
}
`;
